### PR TITLE
Add low level realtime routing table builder

### PR DESCRIPTION
**Pull request by jfim**
Add a new routing table builder for the low level consumer. It's not
tested yet and needs a unit test.

### CODE REVIEW

**Comment by raviarin** [comment]
Please add some unit tests for the two methods ;)

**Comment by jfim** [comment]
Of course :)

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java [hunk 1]
```diff
@@ -42,12 +42,14 @@
       public static class SegmentOnlineOfflineStateModel {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java]: These constants should really be pulled from the state model defintion, but we can leave that for another PR

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    //
+    // This algorithm works as follows:
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    // 2. Ensure that for each partition, we have at most one partition in consuming state
+    // 3. Sort all the segments to be used during assignment in ascending order of replicas
+    // 4. For each segment to be used during assignment, pick a random replica, weighted by the number of segments
+    //    assigned to each replica.
+
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    Map<String, SortedSet<String>> sortedSegmentsByKafkaPartition = new HashMap<String, SortedSet<String>>();
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: where is this initialized?

Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: Also, it helped on the controller side to add a segment holder class that implements Comparable, and sorts segment names by kafka partition and then by offset (we can change offset to time in the final implementation). Perhaps that will be useful here ? (check the validation branch). If so, please introduce a class that holds the lowlevel segment name and extracts components out of it.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+      SortedSet<String> segmentsForPartition = sortedSegmentsByKafkaPartition.get(kafkaPartitionName);
+
+      // Create sorted set if necessary
+      if (segmentsForPartition == null) {
+        segmentsForPartition = new TreeSet<String>(new Comparator<String>() {
+          @Override
+          public int compare(String firstSegment, String secondSegment) {
+            // Sort based on sequence number, falling back on string comparison in case there is an exception
+            try {
+              int firstSegmentSequenceNumber =
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: Isn't it better (and probably more intuitive) to have a LowLevelSegmentNameHolder class?

Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: Sure, we can make that part of another PR. I'd like to avoid scope creep during review time though. :)

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java (hunk 2 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: The routing table builder is initialized by the `RoutingTableBuilderFactory`, I haven't yet added it there (it'll happen when writing the unit tests).

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java (hunk 1 above)
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/CommonConstants.java]: Agreed, we shouldn't have those strings duplicated everywhere.

**Comment by jfim** [comment]
Unit test passing depends on #214

**Comment by mcvsubbu** [comment]
Please use the new classes checked in for low level consumer segment names, thanks.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    // This algorithm works as follows:
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    // 2. Ensure that for each partition, we have at most one partition in consuming state
+    // 3. Sort all the segments to be used during assignment in ascending order of replicas
+    // 4. For each segment to be used during assignment, pick a random replica, weighted by the number of segments
+    //    assigned to each replica.
+
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    Map<String, SortedSet<String>> sortedSegmentsByKafkaPartition = new HashMap<String, SortedSet<String>>();
+    for (String helixPartitionName : externalView.getPartitionSet()) {
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: pls  change the name of the variable from helixPartition to 'segmentName', makes it easier to understand, thkx

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    Map<String, String> lastSegmentInConsumingStateByKafkaPartition = new HashMap<String, String>();
+    for (String kafkaPartition : sortedSegmentsByKafkaPartition.keySet()) {
+      SortedSet<String> sortedSegmentsForKafkaPartition = sortedSegmentsByKafkaPartition.get(kafkaPartition);
+      String lastSegment = sortedSegmentsForKafkaPartition.last();
+
+      // Only keep the segment if all replicas have it in CONSUMING state
+      Map<String, String> helixPartitionState = externalView.getStateMap(lastSegment);
+      for (String externalViewState : helixPartitionState.values()) {
+        // Ignore ERROR state
+        if (externalViewState.equalsIgnoreCase(
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: Can we change the logic to make it explicit check for CONSUMING state? If they are in ERROR or OFFLINE you don't want it anyways, right?

Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: Also, the last segment will always be all in CONSUMING state, but the second-last is the one that may also have consuming state. So, dont you need to check the second-last one?

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java (hunk 2 above)
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: will be really nice if you add a comment near each Map/Set declaration as to what the keys and values are (since everything is a String, it gets a bit confusing).

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+          }
+        });
+
+        sortedSegmentsByKafkaPartition.put(kafkaPartitionName, segmentsForPartition);
+      }
+
+      segmentsForPartition.add(helixPartitionName);
+    }
+
+    // 2. Ensure that for each partition, we have at most one partition in consuming state
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: s/at most one partition/at most one segment/

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        String segment = segmentAndValidReplicaSet.getKey();
+        Set<String> validReplicaSet = segmentAndValidReplicaSet.getValue();
+
+        String replica = pickWeightedRandomReplica(validReplicaSet, instanceToSegmentSetMap);
+        if (replica != null) {
+          Set<String> segmentsForInstance = instanceToSegmentSetMap.get(replica);
+
+          if (segmentsForInstance == null) {
+            segmentsForInstance = new HashSet<String>();
+            instanceToSegmentSetMap.put(replica, segmentsForInstance);
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: should be outside the parenthesis?

Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: ah no, my bad

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java (hunk 2 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: All the variable names follow the `xByY` convention I believe (eg. key is Y, value is X).

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    // Find maximum segment count assigned to a replica
+    String[] replicas = validReplicaSet.toArray(new String[validReplicaSet.size()]);
+    int[] replicaSegmentCounts = new int[validReplicaSet.size()];
+
+    int maxSegmentCount = 0;
+    for (int i = 0; i < replicas.length; i++) {
+      String replica = replicas[i];
+      int replicaSegmentCount = instanceToSegmentSetMap.get(replica).size();
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: wont you get an NPE here initially? since instanceToSegmentSetMap.get(replica) is null to start with?

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java (hunk 5 above)
Comment (jfim) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: No, only the last segment is one for which the consuming state is valid. This is used to store whether the last segment is a segment for which the `CONSUMING` state is valid or not. Any other segment should be in `ONLINE` state to be routable.

File: pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+      if (segmentsForPartition == null) {
+        segmentsForPartition = new TreeSet<SegmentName>();
+
+        sortedSegmentsByKafkaPartition.put(kafkaPartitionName, segmentsForPartition);
+      }
+
+      segmentsForPartition.add(segmentName);
+    }
+
+    // 2. Ensure that for each partition, we have at most one partition in consuming state
```
Comment (mcvsubbu) [inline pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java]: The word partition is overloaded in this comment and a bit confusing. I suggest:
"Ensure that for each kafka partition, we have at most one helix partition in consuming state"
Alternatively, you can use 'segment' insetad of 'helix partition' as suggested in an earlier comment

**Comment by mcvsubbu** [comment]
LGTM otherwise
